proposals/p2875.md

```diff
@@ -0,0 +1,475 @@
+indirect call.
+
+Making functions result in unique types means that calls to functions, lambdas,
+and function-like objects have semantics that are more similar and have similar
+efficiency properties.
+
+### Make direct and indirect calls behave uniformly
+
+It would be desirable to remove the difference between direct and indirect
+calls. Unfortunately, that doesn't seem to be practical, for a number of
+reasons:
+
+-   Indirect calls need to be transformed into another call expression. We need
+    a foundation for our semantics at some point, where we make an actual
+    function call rather than transforming one function call into another.
+-   We have chosen that
+    [interfaces are the only static open extension mechanism](/docs/project/principles/static_open_extension.md).
+    This means that any general mechanism we provide to overload function calls
+    must fit within the boundaries of Carbon's `interface` and `impl`
+    mechanisms.
+-   We could try to support passing compile-time parameters to functions by
+    specifying in the `impl` query whether each argument can be passed at
+    compile time, but any such query will need to be able to fall back to
+    passing each argument at runtime, which can result in an exponential-time
+    search for a matching `impl`.
+
+## Future work
+
+### Overloading
+
+We intend to support function overloading, and although it outside the scope of
+this proposal, it is important to ensure that we have reasonable paths forward
+to support overloading within this framework.
+
+An overloaded function has multiple different signatures -- sets of implicit and
+explicit parameters -- that will be checked against the provided arguments. In
+this proposal, that means that a set of overloaded functions introduces a single
+function type that supports being called in different ways.
+
+Depending on how we approach overloading, overload selection might be done by
+checking each candidate in turn until we find one that matches, or checking them
+all and somehow selecting the best match, and our current intent is to use the
+former approach. Translated into this proposal, one important consequence is
+that the function type corresponding to the set of overloaded functions has
+multiple parameterized implementations of the `Call(...)` interface, and the
+implementation selected for a particular call will need to follow whatever rules
+the overloading mechanism picks. For example, if function overloading picks the
+first matching function, then given the following, using placeholder syntax:
+
+```
+overloaded fn Abs[T:! Unsigned](x: T) -> T { return x; }
+overloaded fn Abs[T:! Floating](x: T) -> T { return x < 0 ? -x : x }
+```
+
+we would synthesize implementations that behave as follows:
+
+```
+match_first {
+  impl forall [T:! Unsigned] Abs as Call(T) where .Result = T { ... }
+  impl forall [T:! Floating] Abs as Call(T) where .Result = T { ... }
```

Comment:
```suggestion
  impl forall [T:! Unsigned] Abs as Call((T,)) where .Result = T { ... }
  impl forall [T:! Floating] Abs as Call((T,)) where .Result = T { ... }
```

In the current design you've chosen to use tuples instead of variadics, so this should be passing a one-length tuple, right?